share/patch/patch_db_zonemaster_backend_ver_11.0.0.pl

```diff
@@ -0,0 +1,232 @@
+use strict;
+use warnings;
+
+use JSON::PP;
+use Try::Tiny;
+
+use Zonemaster::Backend::Config;
+
+my $config = Zonemaster::Backend::Config->load_config();
+
+my %patch = (
+    mysql       => \&patch_db_mysql,
+    postgresql  => \&patch_db_postgresql,
+    sqlite      => \&patch_db_sqlite,
+);
+
+my $db_engine = $config->DB_engine;
+print "engine: $db_engine\n";
+
+if ( $db_engine =~ /^(MySQL|PostgreSQL|SQLite)$/ ) {
+    $patch{ lc $db_engine }();
+}
+else {
+    die "Unknown database engine configured: $db_engine\n";
+}
+
+sub _update_data {
+    my ( $dbh ) = @_;
+
+    my $json = JSON::PP->new->allow_blessed->convert_blessed->canonical;
+
+    my ( $row_total ) = $dbh->selectrow_array( 'SELECT count(*) FROM test_results' );
+    print "count: $row_total\n";
+
+    # depending on the resources available to select all data in database
+    # update $row_count to your needs
+    my $row_count = 50000;
+    my $row_done = 0;
+    while ( $row_done < $row_total ) {
+        print "row_done/row_total: $row_done / $row_total\n";
+        my $sth1 = $dbh->prepare( 'SELECT hash_id, results FROM test_results ORDER BY id ASC LIMIT ?,?' );
+        $sth1->execute( $row_done, $row_count );
+        while ( my $row = $sth1->fetchrow_arrayref ) {
+            my ( $hash_id, $results ) = @$row;
+
+            next unless $results;
+
+            my @records;
+            my $entries = $json->decode( $results );
+
+            foreach my $m ( @$entries ) {
+                my $r = [
+                    $hash_id,
+                    $m->{level},
+                    $m->{module},
+                    $m->{testcase} // '',
+                    $m->{tag},
+                    $m->{timestamp},
+                    $json->encode( $m->{args} // {} ),
+                ];
+
+                push @records, $r;
+            }
+
+            my $query_values = join ", ", ("(?, ?, ?, ?, ?, ?, ?)") x @records;
+            my $query = "INSERT INTO result_entries (hash_id, level, module, testcase, tag, timestamp, args) VALUES $query_values";
+            my $sth = $dbh->prepare( $query );
+            $sth = $sth->execute( map { @$_ } @records );
+
+            $dbh->do( "UPDATE test_results SET results = NULL WHERE hash_id = ?", undef, $hash_id );
+        }
+
+        $row_done += $row_count;
+    }
+}
+
+sub patch_db_mysql {
+    use Zonemaster::Backend::DB::MySQL;
+
+    my $db = Zonemaster::Backend::DB::MySQL->from_config( $config );
+    my $dbh = $db->dbh;
+
+    $dbh->{AutoCommit} = 0;
+
+    try {
+        $dbh->do(
+            "CREATE TABLE IF NOT EXISTS result_entries (
+                id integer AUTO_INCREMENT PRIMARY KEY,
+                hash_id VARCHAR(16) not null,
```

Comment:
I didn't think about it. This could work. We could also remove the `id` field and make `hash_id` the PRIMARY KEY in `test_results`. Regarding performance gain, [this could depend on the database engine](https://stackoverflow.com/questions/39388912/does-adding-a-foreign-key-to-an-indexed-column-boost-performance).
I'll provide an update with this suggestion.

Reply:
I believe joins are faster when using integer keys so it might be a good idea to keep the id field. I guess CHAR(16) is 128 bits, and maybe DBMSs would be able to optimize such joins, but I believe it's best to keep primary keys at 64 bits or smaller.

Reply:
I've added the foreign key referencing the `hash_id` from table `test_results`, and removed the`id` field.